src/main/java/org/jboss/pnc/common/http/PNCHttpClient.java

```diff
@@ -0,0 +1,181 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.pnc.common.http;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import dev.failsafe.Failsafe;
+import dev.failsafe.RetryPolicy;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.jboss.pnc.api.dto.Request;
+
+import java.io.IOException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.function.Supplier;
+
+import static org.jboss.pnc.api.constants.HttpHeaders.AUTHORIZATION_STRING;
+
+/**
+ * Wrapper around {@link HttpClient} that can handle PNC {@link Request}s.
+ */
+@Slf4j
+public class PNCHttpClient {
+    private final HttpClient client;
+    private final ObjectMapper objectMapper;
+    private RetryPolicy<HttpResponse<String>> retryPolicy;
+    @Setter
+    private Supplier<String> tokenSupplier;
+
+    public PNCHttpClient(RetryConfig retryConfig) {
+        this(new ObjectMapper(), retryConfig);
+    }
+
+    public PNCHttpClient(ObjectMapper objectMapper, RetryConfig retryConfig) {
+        this.client = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build();
+        this.objectMapper = objectMapper;
+
+        retryPolicy = RetryPolicy.<HttpResponse<String>> builder()
+                .withBackoff(retryConfig.backoffInitialDelay(), retryConfig.backoffMaxDelay(), ChronoUnit.SECONDS)
+                .withMaxRetries(retryConfig.maxRetries())
+                .withMaxDuration(Duration.of(retryConfig.maxDuration(), ChronoUnit.SECONDS))
+                .onSuccess(e -> log.debug("Request successfully sent, response: {}", e.getResult().statusCode()))
+                .onRetry(
+                        e -> log.debug(
+                                "Retrying (attempt #{}), last response was.",
+                                e.getAttemptCount(),
+                                e.getLastException()))
+                .onFailure(e -> log.warn("Request couldn't be sent.", e.getException()))
+                .build();
+    }
+
+    /**
+     * Tries to send the request once, reporting any error immediately.
+     *
+     * @param request Request to be sent.
+     * @throws PNCHttpClientException When there is error sending the request.
+     */
+    public void trySendRequest(Request request) {
```

Comment:
it's in javadoc, but sure

Reply:
I don't see a reason why it couldn't be there. If I don't have javadoc downloaded, it would give me at least some indication how that method behaves or prompt me to download the docs.

Reply:
Because it's runtime exception and it could be misleading. I don't think it's a good practice to to include runtime exception in method signature.
On the other hand, it's good practice to read javadoc.

Reply:
Internet and Java devs agree with you.